st2actions/st2actions/policies/concurrency.py

```diff
@@ -0,0 +1,94 @@
+
+    def _apply(self, target):
+        # Get the count of scheduled instances of the action.
+        scheduled = action_access.LiveAction.count(
+            action=target.action, status=action_constants.LIVEACTION_STATUS_SCHEDULED)
+
+        # Get the count of running instances of the action.
+        running = action_access.LiveAction.count(
+            action=target.action, status=action_constants.LIVEACTION_STATUS_RUNNING)
+
+        count = scheduled + running
+
+        # Mark the execution as scheduled if threshold is not reached or delayed otherwise.
+        if count < self.threshold:
+            LOG.debug('There are %s instances of %s in scheduled or running status. '
+                      'Threshold of %s is not reached. Action execution will be scheduled.',
+                      count, target.action, self._policy_ref)
+            status = action_constants.LIVEACTION_STATUS_SCHEDULED
+        else:
+            LOG.debug('There are %s instances of %s in scheduled or running status. '
+                      'Threshold of %s is reached. Action execution will be delayed.',
+                      count, target.action, self._policy_ref)
+            status = action_constants.LIVEACTION_STATUS_DELAYED
+
+        # Update the status in the database but do not publish.
+        target = action_utils.update_liveaction_status(
+            status=status,
+            liveaction_id=target.id,
+            publish=False)
+
+        action_execution_db = executions.update_execution(target)
+
+        extra = {'action_execution_db': action_execution_db, 'liveaction_db': target}
+        LOG.audit('The status of action execution (id=%s) / live action (id=%s) '
+                  'is changed from %s to %s.', action_execution_db.id, target.id,
+                  target.status, status, extra=extra)
+
+        LOG.info('The status of action execution (id=%s) / live action (id=%s) '
+                 'is changed from %s to %s.', action_execution_db.id, target.id,
+                 target.status, status)
+
+        return target
+
+    def apply(self, target):
+        # Exit if target not in schedulable state.
+        if target.status != action_constants.LIVEACTION_STATUS_REQUESTED:
+            LOG.debug('The live action is not schedulable. %s', target)
+            return target
+
+        # Warn users that the coordination service is not configured.
+        if not coordination.configured():
+            LOG.warn('Coordination service is not configured. Policy enforcement is best effort.')
+
+        # Acquire a distributed lock before querying the database to make sure that only one
+        # scheduler is scheduling execution for this action. Even if the coordination service
+        # is not configured, the fake driver using zake or the file driver can still acquire
+        # a lock for the local process or server respectively.
+        with self.coordinator.get_lock(target.action):
```

Comment:
seems like a very general lock. Can you append a policy name so something like `concurreny/action_ref` so that all policies do not get stuck behind the same lock?